### PR TITLE
SI-10023: third iteration using member.keyword and special handling for parameterless methods

### PR DESCRIPTION
**Pull request by esarbe**
I guess that the strings _probably_ shouldn't be hard coded. What would be a more appropriate way to do it?

### CODE REVIEW

**Comment by som-snytt** [comment]
Probably. https://github.com/scala/scala/pull/5222

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -2490,7 +2490,7 @@ trait Types
     //TODO this may be generalised so that the only constraint is dependencies are acyclic
     def approximate: MethodType = MethodType(params, resultApprox)
 
-    override def safeToString = paramString(this) + resultType
+    override def safeToString = paramString(this) + ": " + resultType
```
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Types.scala]: I would recommend restricting the changes for this PR to the interpreter, as this method in reflect should stay the same.

Comment (esarbe) [reply src/reflect/scala/reflect/internal/Types.scala]: Thanks for the feedback. I thought that it was too easy. ;)

**Comment by esarbe** [comment]
This version has the problem of parameterless methods being printed as 

```
scala> def one = 1
def one(): Int
```

instead of (as it currently is)

```
scala> def one = 1
one: Int
```

although one could argue that the current output is also not consistent, since according to
http://www.scala-lang.org/files/archive/spec/2.11/03-types.html#method-types
it should actually be

```
scala> def one = 1
one: => Int
```

So, what do do? I'm loath to match on the type. Suggestions? I'm sure I shouldn't go as far as to introduce a `NullaryDefHandler`...

**Comment by SethTisue** [comment]
I believe you're trying to match the language syntax, not the method type notation in the spec. Method types are a spec thing, not a language thing. So `def one: Int` (not `def one(): Int`) seems correct to me.

**Comment by som-snytt** [comment]
If the feature tends toward minimally two modes, beginner and expert, or verbose and terse, I'd go for "type" mode myself:

```
scala> :type val x = 42
Int

scala> :type def f = 42
=> Int

scala> def f = 42
f: => Int                            // conforms with :type output, no val or def keyword
```

Another possibility is for verbose mode to be "source mode" or "print-tab mode":

```
scala> def f = 42
def f: Int = 42

scala> def f = 42
def f = 42 : Int
```

print-tab puts the ascribed type behind a line comment, but the idea is to highlight that this is the inferred type. It's easy to cut/paste the prefix of the output that excludes it, though maybe there's no use case for that, unless:

```
scala> def g(implicit s: String) = s
g: (implicit s: String)String  // old-style output

scala> def f(implicit s: String) = g
def f(implicit s: String) = g(s) : String    // may want to cut/paste expanded version?
```

Another idea is to have a special color for inferred types, and put them in usual source order.

**Comment by esarbe** [comment]
@som-snytt Do you think I should create another SI for the beginner/expert switch? Or could this be addressed in the context of SI-10024?

**Comment by esarbe** [comment]
This version also handles parameterless methods. I don't like the if/else, so I'd be glad for any pointer to get it done without.

**Pull request closed without merging**